### PR TITLE
Add mixed precision support to Transformer

### PR DESCRIPTION
**Pull request by reedwm**
To test convergence, I ran the following command on a DGX-1:

```
python transformer_main.py --num_gpus=8 --distribution_strategy=mirrored --data_dir ~/transformer_data2 --vocab_file ~/transformer_data2/vocab.ende.32768 --bleu_source ~/transformer_data2/newstest2014.en --bleu_ref ~/transformer_data2/newstest2014.de --param_set=base --batch_size=32768 --train_steps=300000 --steps_between_evals=20000 --log_steps=1000 --static_batch --max_length=64  --model_dir ~/transformer_model_dir_accuracy_fp16 --clean --dtype=fp16
```

I ran the convergence test on a slightly older version of the PR (67a4bdfc689a7e530663fd40614cccf25bd2df24) before doing some cleanup.

The results: BLEU: uncased=27.7248412371, cased=27.1868109703.

EDIT:

And the fp32 results: BLEU: uncased=27.6445806026, cased=27.1022170782
So fp16 does not harm accuracy with the command I tried.

### CODE REVIEW

**Comment by tfboyd** [comment]
Removing myself as others can review this one much better. I will review when the benchmark is added.

File: official/transformer/v2/misc.py [hunk 1]
```diff
@@ -68,7 +68,8 @@ def define_transformer_flags():
       intra_op=False,
       synthetic_data=True,
       max_train_steps=False,
-      dtype=False,
+      dtype=True,
+      loss_scale=True,
```
Comment (azaks2) [inline official/transformer/v2/misc.py]: why true by default?

File: official/transformer/v2/transformer.py [hunk 2]
```diff
@@ -299,12 +311,14 @@ def build(self, input_shape):
         "layer_norm_scale",
         shape=[self.hidden_size],
         dtype="float32",
-        initializer=tf.ones_initializer())
+        initializer=tf.ones_initializer(),
+        experimental_autocast=False)
```
Comment (azaks2) [inline official/transformer/v2/transformer.py]: add a comment about experimental_autocast

File: official/transformer/v2/transformer.py [hunk 3]
```diff
@@ -313,10 +327,13 @@ def get_config(self):
     }
 
   def call(self, x, epsilon=1e-6):
+    orig_dtype = x.dtype
```
Comment (azaks2) [inline official/transformer/v2/transformer.py]: rename to input_dtype?

File: official/transformer/v2/transformer.py (hunk 2 above)
Comment (reedwm) [reply official/transformer/v2/transformer.py]: Done

File: official/transformer/v2/transformer.py (hunk 3 above)
Comment (reedwm) [reply official/transformer/v2/transformer.py]: Done

File: official/transformer/v2/misc.py (hunk 1 above)
Comment (reedwm) [reply official/transformer/v2/misc.py]: "True" means to have the --dtype and --loss_scale flags, which I am now exposing.

File: official/transformer/v2/attention_layer.py [hunk 4]
```diff
@@ -21,6 +21,14 @@
 import tensorflow as tf
 
 
+def _float32_softmax(logits, name=None):
+  """Computes a softmax activation in float32."""
+  input_dtype = logits.dtype
+  logits = tf.cast(logits, tf.float32)
+  output = tf.nn.softmax(logits, name=name)
```
Comment (isaprykin) [inline official/transformer/v2/attention_layer.py]: Would it be appropriate to add some documentation for why do we need to re-cast to fp32 right before softmax and then back?

Comment (reedwm) [reply official/transformer/v2/attention_layer.py]: Done. Updated the docstring.

**Comment by venuswu** [comment]
Do you mean you ran under v2 ? I have ran fp16 under v1, but it diverged ?

**Comment by venuswu** [comment]
> To test convergence, I ran the following command on a DGX-1:
> 
> ```
> python transformer_main.py --num_gpus=8 --distribution_strategy=mirrored --data_dir ~/transformer_data2 --vocab_file ~/transformer_data2/vocab.ende.32768 --bleu_source ~/transformer_data2/newstest2014.en --bleu_ref ~/transformer_data2/newstest2014.de --param_set=base --batch_size=32768 --train_steps=300000 --steps_between_evals=20000 --log_steps=1000 --static_batch --max_length=64  --model_dir ~/transformer_model_dir_accuracy_fp16 --clean --dtype=fp16
> ```
> 
> I ran the convergence test on a slightly older version of the PR ([67a4bdf](https://github.com/tensorflow/models/commit/67a4bdfc689a7e530663fd40614cccf25bd2df24)) before doing some cleanup.
> 
> The results: BLEU: uncased=27.7248412371, cased=27.1868109703.
> 
> EDIT:
> 
> And the fp32 results: BLEU: uncased=27.6445806026, cased=27.1022170782
> So fp16 does not harm accuracy with the command I tried.

Do you mean you ran under v2 ? I have ran fp16 under v1, but it diverged ?
@reedwm

**Comment by reedwm** [comment]
Yes, I ran with the v2 version of Transformer at official/transformer/v2/transformer_main.py, with TensorFlow 2. #6756 added support to the v1 version of Transformer.

**Comment by venuswu** [comment]
I ran v1 version of Transformer with fp16 mixed-precesion supported, but it deverged. @reedwm

**Comment by venuswu** [comment]
Why do you force the fp32 op in the above commit? Can the model deverge without these fp32 op?

**Comment by reedwm** [comment]
With #6756, the graph is built in float32, then a graph rewrite changes the graph to use float16.

I am not sure why it diverged. For the v1 version, the following flags got use a BLEU of 27.69:

```
--batch_size=32768 --bleu_ref=/data/newstest2014/newstest2014.de --bleu_source=/data/newstest2014/newstest2014.en --data_dir=/data/wmt32k-en2de-official --dtype=fp16 --hooks=ExamplesPerSecondHook --model_dir=/workspace/benchmarks/perfzero/workspace/output/2019-08-09-13-03-35-870157/benchmark_graph_fp16_8_gpu --num_gpus=8 --param_set=base --steps_between_evals=5000 --train_steps=100000 --vocab_file=/data/wmt32k-en2de-official/vocab.ende.32768
```

Try those flags. You'll have to modify the file paths to point to appropriate locations on your machine. Note that unlike the flags in the PR description, these flags are for the V1 version of transformer, not the V2 version

**Comment by venuswu** [comment]
the batchsize is so big, what is the memory of you gpu, Does such a big batchsize  lead to OOM？
@reedwm

**Comment by venuswu** [comment]
I ran with gpu_num=8 under tf1.14, but it seems to be very slow, 0.4steps/s. @reedwm

**Comment by reedwm** [comment]
@venuswu, we run on 8 V100 GPUs with 16GB of memory each. You should lower the batch size if you have less memory. Also, `--num_gpus=8` is only useful if you have 8 GPUs. If you only have 1 GPU, set num_gpus to 1 instead.

**Comment by tfboyd** [comment]
@venuswu  A small warning, at smaller batch sizes transformer does not converge to as high an accuracy in our testing.  The testing was not extensive; but I recall that being a problem when we move the model to the TF 2.0 style.  We had a vareity of issues so it may have been a red herring; but I do recall a similar finding in the Tensor2Tensor github issues list as well as maybe a small paper written on the subject.  I did not run Transformer V1 FP16 very often, if you end up using the v2 (TensorFlow 2.0), I can share a bunch of data that might be helpful.  Good luck.

**Comment by venuswu** [comment]
How can I get the graph for tensorboard after mixed precision? @reedwm @tfboyd

**Comment by reedwm** [comment]
Looks like you figured it out in #7480. We can discuss the problem in that issue.

**Comment by venuswu** [comment]
> With #6756, the graph is built in float32, then a graph rewrite changes the graph to use float16.
> 
> I am not sure why it diverged. For the v1 version, the following flags got use a BLEU of 27.69:
> 
> ```
> --batch_size=32768 --bleu_ref=/data/newstest2014/newstest2014.de --bleu_source=/data/newstest2014/newstest2014.en --data_dir=/data/wmt32k-en2de-official --dtype=fp16 --hooks=ExamplesPerSecondHook --model_dir=/workspace/benchmarks/perfzero/workspace/output/2019-08-09-13-03-35-870157/benchmark_graph_fp16_8_gpu --num_gpus=8 --param_set=base --steps_between_evals=5000 --train_steps=100000 --vocab_file=/data/wmt32k-en2de-official/vocab.ende.32768
> ```
> 
> Try those flags. You'll have to modify the file paths to point to appropriate locations on your machine. Note that unlike the flags in the PR description, these flags are for the V1 version of transformer, not the V2 version

What's the version of tensorflow you run ?
